include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+{
+#ifdef PARAM3_ALWAYS_SAME_VALUE
+	return PARAM3_VALUE;
+#else
+	return textureOrRGBconst(uvec3(data.data[1].zw, data.data[2].x), texPresence);
+#endif
+}
+vec3 bsdf_data_getParam4(in bsdf_data_t data, in bool texPresence)
+{
+#ifdef PARAM4_ALWAYS_SAME_VALUE
+	return PARAM4_VALUE;
+#else
+	return textureOrRGBconst(data.data[2].yzw, texPresence);
+#endif
+}
+
+bvec4 instr_getTexPresence(in instr_t i)
+{
+	return bvec4(
+		instr_get1stParamTexPresence(i),
+		instr_get2ndParamTexPresence(i),
+		instr_get3rdParamTexPresence(i),
+		instr_get4thParamTexPresence(i)
+	);
+}
+params_t instr_getParameters(in instr_t i, in bsdf_data_t data)
+{
+	params_t p;
+	bvec4 presence = instr_getTexPresence(i);
+	//speculatively always read RGB
+	p[0] = bsdf_data_getParam1(data, presence.x);
+	p[1] = bsdf_data_getParam2(data, presence.y);
+	p[2] = bsdf_data_getParam3(data, presence.z);
+	p[3] = bsdf_data_getParam4(data, presence.w);
+
+	return p;
+}
+
+//this should thought better
+mat2x3 bsdf_data_decodeIoR(in bsdf_data_t data, in uint op)
+{
+	mat2x3 ior = mat2x3(0.0);
+#ifdef OP_CONDUCTOR
+	if (op==OP_CONDUCTOR) {
+		ior[0] = irr_glsl_decodeRGB19E7(data.data[2].yz);
+		ior[1] = irr_glsl_decodeRGB19E7(uvec2(data.data[2].w,data.data[3].x));
+	} else
+#endif
+#ifdef OP_PLASTIC
+	if (op==OP_PLASTIC) {
+		ior[0] = vec3(uintBitsToFloat(data.data[3].x));
+	} else
+#endif
+#ifdef OP_DIELECTRIC
+	if (op==OP_DIELECTRIC) {
+		ior[0] = vec3(uintBitsToFloat(data.data[3].x));
+	} else
+#endif
+#ifdef OP_THINDIELECTRIC
+		if (op == OP_THINDIELECTRIC) {
+			ior[0] = vec3(uintBitsToFloat(data.data[3].x));
+		}
+		else
+#endif
+#ifdef OP_COATING
+	if (op==OP_COATING) {
+		ior[0] = vec3(unpackHalf2x16(data.data[3].x).y);
+	} else
+#endif
+	{}
+
+	return ior;
+}
+
+void writeReg(in uint n, in float v)
+{
+	registers[n] = floatBitsToUint(v);
+}
+void writeReg(in uint n, in vec2 v)
+{
+	writeReg(n   ,v.x);
+	writeReg(n+1u,v.y);	
+}
+void writeReg(in uint n, in vec3 v)
+{
+	writeReg(n   ,v.x);
+	writeReg(n+1u,v.y);
+	writeReg(n+2u,v.z);
+}
+void writeReg(in uint n, in vec4 v)
+{
+	writeReg(n   ,v.xyz);
+	writeReg(n+3u,v.w);
+}
+float readReg1(in uint n)
+{
+	return uintBitsToFloat( registers[n] );
+}
+vec2 readReg2(in uint n)
+{
+	return vec2(
+		readReg1(n), readReg1(n+1u)
+	);
+}
+vec3 readReg3(in uint n)
+{
+	return vec3(
+		readReg1(n), readReg1(n+1u), readReg1(n+2u)
+	);
+}
+vec4 readReg4(in uint n)
+{
+	return vec4(
+		readReg3(n), readReg1(n+3u)
+	);
+}
+
+mat2x4 instr_fetchSrcRegs(in instr_t i, in uvec3 regs)
+{
+	uvec3 r = regs;
+	mat2x4 srcs;
+	srcs[0] = readReg4(REG_SRC1(r));
+	srcs[1] = readReg4(REG_SRC2(r));
+	return srcs;
+}
+mat2x4 instr_fetchSrcRegs(in instr_t i)
+{
+	uvec3 r = instr_decodeRegisters(i);
+	return instr_fetchSrcRegs(i, r);
+}
+
+void setCurrInteraction(in vec3 N)
+{
+	vec3 campos = irr_glsl_MC_getCamPos();
+	irr_glsl_IsotropicViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, N);
```

Comment:
the stuff that flows out of `campos` and `WorldPos` should be precomputed.... i.e. the normalized `V`

Reply:
you should have `initInteraction` and `updateInteraction` really...